ow_power_system/src/PowerSystemNode.cpp

```diff
@@ -2,319 +2,374 @@
+    return;
   }
 
-  return true;
-}
+  // Set up the time interval & moving average size.
+  m_time_interval = 1 / m_loop_rate_hz;
+  m_moving_average_window = floor(m_joint_states_rate / m_loop_rate_hz);
 
-bool PowerSystemNode::loadSystemConfig()
-{
-  auto system_config_path = ros::package::getPath("ow_power_system")
-    + "/config/system.cfg";
-  auto system_config = ConfigMap(system_config_path);
-  m_initial_power = system_config.getDouble("initial_power");
-  m_initial_voltage = system_config.getDouble("initial_voltage");
-  m_initial_temperature = system_config.getDouble("initial_temperature");
-  m_base_voltage = system_config.getDouble("base_voltage");
-  m_voltage_range = system_config.getDouble("voltage_range");
-  m_min_temperature = system_config.getDouble("min_temperature");
-  m_max_temperature = system_config.getDouble("max_temperature");
-  m_battery_lifetime = system_config.getDouble("battery_lifetime");
-  m_efficiency = system_config.getDouble("efficiency");
-  m_temperature_dist = uniform_real_distribution<double>(m_min_temperature,
-                                                         m_max_temperature);
-  m_baseline_wattage = system_config.getDouble("baseline_power");
-  m_max_gsap_input_watts = system_config.getDouble("max_gsap_power_input");
-  m_gsap_rate_hz = system_config.getDouble("gsap_rate");
-  m_profile_increment = system_config.getInt32("profile_increment");
-  m_moving_average_window = system_config.getInt32("power_average_size");
-  m_spinner_threads = system_config.getInt32("spinner_threads");
-  return true;
-}
+  // Set up the moving average vector.
+  m_power_values.resize(m_moving_average_window, 0.0);
 
-PrognoserVector PowerSystemNode::loadPowerProfile(const string& path_name, string custom_file)
-{
-  ifstream file(path_name);
+  initTopics();
 
-  if (file.fail())
+  // Initialize EoD_events and previous_times.
+  for (int i = 0; i < NUM_MODELS; i++)
   {
-    ROS_WARN_STREAM("Could not find a custom file using the path '"
-                          << path_name << "'. Deactivate fault and try again.");
-    return PrognoserVector();
+    m_power_models[i].previous_time = 0;
+    m_EoD_events[i].remaining_useful_life = m_max_horizon_secs;
+    m_EoD_events[i].state_of_charge = m_initial_soc;
+    m_EoD_events[i].battery_temperature = m_initial_temperature;
   }
 
-  // Skip header line
-  file.ignore(numeric_limits<streamsize>::max(), '\n');
-
-  auto now = system_clock::now();
+  // Construct the prediction handlers.
+  auto handlers = std::array<std::unique_ptr<PredictionHandler>, NUM_MODELS>();
+  for (int i = 0; i < NUM_MODELS; i++)
+  {
+    handlers[i] = std::make_unique<PredictionHandler>(
+                    m_power_models[i].bus, m_power_models[i].name
+    );
+  }
 
-  PrognoserVector result;
 
-  // Line number starts at 2 instead of 1 because the first line is the header.
-  int line_number = 2;
+  std::vector<PCOE::AsyncPrognoser> prognosers;
```

Comment:
Out of curiosity, what would be gained from this change? I've always heard to use vectors instead of arrays if possible.

I'm getting an error when trying to initialize an empty array of ``PCOE::AsyncPrognoser``s, which I could probably figure out, but I'm not sure this kind of change is high priority in the first place.

Reply:
I think that's an over-simplified rule. `std::array` makes sense so long as you know the size at runtime, which I think has the potential to lend you some memory optimization, but in this case the vector is so small it really makes no difference. And I have to agree with you, I don't see how you could call the `build` method from the GSAP API in a way to populates a static-length array.

I think you did the right thing here, so I'll resolve this comment.